cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	return strconv.ParseUint(token, 10, 64)
+}
+
+func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
+	line = strings.TrimSpace(line)
+
+	// ignore comments and empty lines
+	if strings.HasPrefix(line, "#") || line == "" {
+		return nil
+	}
+
+	// regular line
+	tokens := strings.Fields(line)
+
+	// fish out syscall
+	secSyscall, err := seccomp.GetSyscallFromName(tokens[0])
+	if err != nil {
+		// FIXME: add structed error to libseccomp-golang
+		if err.Error() == "could not resolve name to syscall" {
+			return nil
+		}
+		return fmt.Errorf("cannot resolve name: %s", err)
+	}
+
+	var conds []seccomp.ScmpCondition
+	if err != nil {
+		return fmt.Errorf("cannot create new filter: %s", err)
+	}
+
+	for pos, arg := range tokens[1:] {
+		var cmpOp seccomp.ScmpCompareOp
+		var value uint64
+		var err error
+
+		if pos >= ScArgsMaxlength {
+			return fmt.Errorf("too many tokens (%d) in line %q", pos, line)
+		}
+
+		if arg == "-" {
+			continue
+		}
+
+		if strings.HasPrefix(arg, ">=") {
+			cmpOp = seccomp.CompareGreaterEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "<=") {
+			cmpOp = seccomp.CompareLessOrEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "!") {
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// FIXME: why do we need this fallback?
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	switch arch.UbuntuArchitecture() {
+	case "amd64":
+		return secFilter.AddArch(seccomp.ArchX86)
+	case "arm64":
+		return secFilter.AddArch(seccomp.ArchARM)
+	case "ppc64el":
+		return secFilter.AddArch(seccomp.ArchPPC)
+	}
+	return nil
```

Comment:
I'm curious to learn more about this. If the userspace is i386 but the kernel is amd64, why do we need add the amd64 syscalls in the seccomp then? Is that for the case that someone does `apt install snapd:i386` on an amd64 system? But even with that we would only be able to install i386 snaps. I'm sure I'm missing something and I'm curious to learn what.